testing/src/main/java/org/oppia/android/testing/junit/OppiaParameterizedTestRunner.kt

```diff
@@ -0,0 +1,303 @@
+        }
+      }
+      return@map ParameterizedMethod(method.name, allValues, fields)
+    }.associateBy { it.methodName }
+  }
+
+  private fun fetchParameterizedFields(): List<Field> {
+    return testClass.declaredFields.mapNotNull { field ->
+      field.getDeclaredAnnotation(Parameter::class.java)?.let { field }
+    }
+  }
+
+  private fun fetchParameterizedMethodDeclarations(): List<ParameterizedMethodDeclaration> {
+    return testClass.declaredMethods.mapNotNull { method ->
+      method.getDeclaredAnnotationsByType(Iteration::class.java).map { parameters ->
+        parameters.name to parameters.keyValuePairs.toList()
+      }.takeIf { it.isNotEmpty() }?.let { rawValues ->
+        val groupedValues = rawValues.groupBy({ it.first }, { it.second })
+        // Verify there are no duplicate iteration names.
+        groupedValues.forEach { (iterationName, iterations) ->
+          check(iterations.size == 1) {
+            "Encountered duplicate iteration name: $iterationName in method ${method.name}"
+          }
+        }
+        val mappedValues = groupedValues.mapValues { (_, iterations) -> iterations.first() }
+        ParameterizedMethodDeclaration(method, mappedValues)
+      }
+    }
+  }
+
+  /**
+   * Defines a parameter that may have an injected value that comes from per-test [Iteration]
+   * definitions.
+   *
+   * It's recommended to make such fields 'lateinit var', and they must be public.
+   *
+   * The type of the parameter field will define how [Iteration]-defined values are parsed. The
+   * current list of supported types:
+   * - [String]s
+   * - [Boolean]s
+   * - [Int]s
+   * - [Long]s
+   * - [Float]s
+   * - [Double]s
+   */
+  @Target(AnnotationTarget.FIELD) annotation class Parameter
+
+  /**
+   * Specifies that a method in a test that uses a [OppiaParameterizedTestRunner] runner should be
+   * run multiple times for each [Iteration] specified in the [value] iterations list.
+   *
+   * See the KDoc for the runner for example code.
+   */
+  @Target(AnnotationTarget.FUNCTION) annotation class RunParameterized(vararg val value: Iteration)
+
+  // TODO(#4120): Migrate to Kotlin @Repeatable once Kotlin 1.6 is used (see:
+  //  https://youtrack.jetbrains.com/issue/KT-12794).
+  /**
+   * Defines an iteration to run as part of a [RunParameterized]-annotated test method.
+   *
+   * See the KDoc for the runner for example code.
+   *
+   * @property name the name of the iteration (this should be short, but meaningful since it's
+   *     appended to the test name)
+   * @property keyValuePairs an array of strings of the form "key=value" where 'key' is the name of
+   *     a [Parameter]-annotated field and 'value' is a stringified conforming value based on the
+   *     type of that field (incompatible values will result in test failures)
+   */
+  @Repeatable(RunParameterized::class)
+  @Target(AnnotationTarget.FUNCTION)
+  annotation class Iteration(val name: String, vararg val keyValuePairs: String)
```

Comment:
How are we handling the use case where if any single test gets failed in the iteration or is it handled by `AndroidJUnit4ClassRunner` only?

Reply:
If one of the iterations fail, it'll show up as the test's name following by a dash and the name of the iteration. You can actually directly run that iteration with Bazel  (and presumably Gradle, but I didn't try).

I'm not sure that I understand the second part of your question--what do you mean by a test being handled by ``AndroidJUnit4ClassRunner`` only?

Reply:
I was guessing `AndroidJUnit4ClassRunner` is somehow helping us here in the iteration fail part.